File: src/main/java/Year2022/Month03/Day28/Part2/GenericVariation.java
package Year2022.Month03.Day28.Part2;

import java.util.ArrayList;
import java.util.List;

/*
@Machine: RedmiBook Pro 15 
@Author: Barry
@Time: 2022/3/28 22:26 周一
@Description: 型变
*/
public class GenericVariation {
    public static void main(String[] args) {
        List<? extends Animal> animals = new ArrayList<>();
        // List<Cat> 变成了 List<? extends Animal>的子类型，即animals变成了可以放入任何Animal及其子类的List
        animals = new ArrayList<Cat>();
    }
}

File: src/main/kotlin/Year2022/Month03/Day28/GenericVariation.kt
package Year2022.Month03.Day28

import Year2022.Month03.Day28.Part2.Animal
import Year2022.Month03.Day28.Part2.Cat

/* 
@Machine: RedmiBook Pro 15 
@Author: Barry
@Time: 2022/3/28 23:17 周一
@Description: 协变
*/
class GenericVariation

fun main() {
    var animals : List<Animal> = ArrayList()
    val cats = ArrayList<Cat>()
    animals = cats
    // 没报错，其实在Kotlin中List的源码使用了out关键字，就相当于Java的上界通配符。
    // 分析源码发现@UnsafeVariance目的为了打破out使用的限制，否则编译报错。
}